src/docs/app-monitoring/user-defined-monitoring.md

```diff
@@ -0,0 +1,221 @@
+
+### Counter
+
+A counter is a cumulative metric that represents a single monotonically increasing counter whose value can only increase or be reset to zero on restart. For example, you can use a counter to represent the number of requests served, tasks completed, or errors.​
+
+https://prometheus.io/docs/concepts/metric_types/#counter
+
+### Gauge
+
+A gauge is a metric that represents a single numerical value that can arbitrarily go up and down.​
+
+Gauges are typically used for measured values like temperatures or current memory usage, but also "counts" that can go up and down, like the number of concurrent requests.​
+
+https://prometheus.io/docs/concepts/metric_types/#gauge
+
+### Histogram and Summaries
+
+**Histograms** and **summaries** both sample observations, typically request durations or response sizes. They track the number of observations and the sum of the observed values, allowing you to calculate the average of the observed values.​
+
+You can use both summaries and histograms to calculate so-called `φ-quantiles`, where 0 ≤ φ ≤ 1. The `φ-quantile` is the observation value that ranks at number `φ*N among the N observations`.
+
+Examples for φ-quantiles:
+
+- The `0.5-quantile` is known as the `median - 50th percentaile`.
+- The `0.95-quantile` is the `95th percentile`.​
+
+The essential difference between summaries and histograms is that summaries calculate streaming `φ-quantiles` on the client side and expose them directly, while histograms expose bucketed observation counts and the calculation of quantiles from the buckets of a histogram happens on the server side using the `histogram_quantile()` function.​
+
+https://prometheus.io/docs/practices/histograms/
+
+## Expose the metrics from your app
+
+Now that your app is generating a metrics endpoint, expose it with a Service.​ The key part is making sure the ports match up with what your library is using.
+
+```yaml
+apiVersion: v1​
+kind: Service​
+metadata:​
+  labels:​
+    app: myapp​
+  name: myapp​
+spec:​
+  ports:​
+  - name: metrics​
+    port: 8000​
+    protocol: TCP​
+    targetPort: 8000​
+  selector:​
+    app: myapp​
+  sessionAffinity: None​
+  type: ClusterIP 
+```
+
+## Create a ServiceMonitor
+
+To use the metrics exposed by your service, you must configure OCP monitoring to scrape metrics from the /metrics endpoint. You can do this using a ServiceMonitor custom resource definition (CRD) that specifies how a service should be monitored.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1​
+kind: ServiceMonitor​
+metadata:​
+  labels:​
+    app: myapp​
+  name: myapp​
+spec:​
+  endpoints:​
+  - interval: 30s​
+    port: metrics​
+    scheme: http​
+  selector:​
+    matchLabels:​
+      app: myapp
+```
+
+![user defined monitoring](../../images/user-defined-monitoring.png)
```

Comment:
curious is there a way to save/bookmark the queries in openshift?

Reply:
Hmmm, unlike sisdig, I think there is no handy way to save your custom queries in OCP console. Two workarounds I can think of;

  [1] Save Prometheus query url into your browser.
  [2] Use grafana-operator

[1] build you query in the OCP web console like the picture in this doc. Then bookmark that URL into your browser.

That url should be Prom's HTTP API query and re-useable, something looks like this:

- https://console.apps.klab.devops.gov.bc.ca/monitoring/query-browser?query0=sum%28response_time_seconds%7Burl%3D%22https%3A%2F%2Fnginx-openshift-bcgov-nagios.apps.klab.devops.gov.bc.ca%2Ftest.txt%22%7D%29+by+%28metric%29

Note: This is KLAB console and `openshift-bcgov-nagios` project. You need to login to KLAB first and then open this url. 

- [Prometheus HTTP API](https://prometheus.io/docs/prometheus/latest/querying/api/)
 
[2] I think this is doable by the users by deploying the grafana-operator in their namespace, and you can save your queries and deploy as a grafana dashboard. But it's more like for cluster-admins, and since sisdig is opened for users, they may prefer sisdig.  

Example is below.  These queries are managed by Argo CCM.
 
- https://grafana-route-openshift-bcgov-grafana.apps.clab.devops.gov.bc.ca/d/XvWDQi9Mz/cluster-capacity-all-namespaces?orgId=1

- [Grafana Operator for Clusters](https://github.com/bcgov-c/platform-gitops-gen/tree/master/roles/grafana-operator#grafana-operator-for-clusters)